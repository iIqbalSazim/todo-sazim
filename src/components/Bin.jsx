import { ActionIcon, Group, Menu } from "@mantine/core";
import { IconTrashFilled } from "@tabler/icons-react";

const Bin = ({ trash, handleRetrieveAllClick, handleEmptyBinClick }) => {
  return (
    <Group position="right" my={"xl"}>
      {trash.length !== 0 ? (
        <Menu width={200} shadow="md" withArrow position="left" offset={0}>
          <Menu.Target>
            <ActionIcon size={"4rem"} color="red.4" variant="transparent">
              <IconTrashFilled stroke="1px" size={"3rem"} />
            </ActionIcon>
          </Menu.Target>

          <Menu.Dropdown>
            <Menu.Item onClick={() => handleRetrieveAllClick()}>
              Retrieve all
            </Menu.Item>

            <Menu.Item onClick={() => handleEmptyBinClick()}>
              Empty Bin
            </Menu.Item>
          </Menu.Dropdown>
        </Menu>
      ) : null}
    </Group>
  );
};

export default Bin;
