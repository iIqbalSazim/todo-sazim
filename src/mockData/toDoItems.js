export const mockToDoItems = [
  {
    id: 1,
    description: "Test task 1",
    priority: "medium",
    created_at: "2023-11-30T09:45:36.302Z",
    due_date: "2023-12-22T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
  {
    id: 2,
    description: "Test task 2",
    priority: "high",
    created_at: "2023-11-27T11:05:37.765Z",
    due_date: "2023-12-30T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
  {
    id: 3,
    description: "Test task 3",
    priority: "low",
    created_at: "2023-11-30T12:33:57.522Z",
    due_date: "2023-11-29T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
];

export const mockToDo = {
  id: 1,
  description: "Task description",
  priority: "high",
  due_date: "2023-12-31T00:00:00.000Z",
};

export const mockTrashTodos = [
  {
    id: 1,
    description: "Test task 1",
  },
  {
    id: 2,
    description: "Test task 2",
  },
];
