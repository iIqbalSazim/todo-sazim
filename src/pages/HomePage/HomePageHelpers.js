export function findIndexWithId(data, id) {
  return data.findIndex((element) => element.id === id);
}

export function formatCompletedAt(dateString) {
  const date = new Date(dateString);

  let currentTime =
    date.toLocaleTimeString().slice(0, 4) + date.toLocaleTimeString().slice(7);
  let currentDate = `${currentTime} | ${date.toDateString()}`;
  return currentDate;
}

export function formatDueDate(dateString) {
  const date = new Date(dateString);
  return date.toDateString().slice(0, 15);
}

export const replaceTabId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-tab-(all|active|completed|target)"/g,
    (match, p1) => `id="mantine-mockedId-tab-${p1}"`
  );
};

export const replaceMenuTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(target)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDropdownTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(dropdown)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDataFocusId = (htmlString) => {
  return htmlString.replace(
    /data-focus-id="mantine-[a-z0-9]+"/g,
    'data-focus-id="mantine-mockedId"'
  );
};

export const replaceAriaLabelledBy = (htmlString) => {
  return htmlString.replace(
    /aria-labelledby="mantine-[a-z0-9]+-target"/g,
    'aria-labelledby="mantine-mockedId-target"'
  );
};

export const replaceAriaControls = (htmlString) => {
  return htmlString.replace(
    /aria-controls="mantine-[a-z0-9]+-dropdown"/g,
    'aria-controls="mantine-mockedId-dropdown"'
  );
};

export const replaceAllRandomGeneratedIds = (htmlString) => {
  let result = replaceTabId(htmlString);
  result = replaceMenuTargetId(result);
  result = replaceDropdownTargetId(result);
  result = replaceDataFocusId(result);
  result = replaceAriaLabelledBy(result);
  result = replaceAriaControls(result);

  return result;
};
