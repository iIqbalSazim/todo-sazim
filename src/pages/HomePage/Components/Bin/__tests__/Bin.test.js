import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { mockTrashTodos } from "../../../../../mockData/toDoItems";
import { replaceAllRandomGeneratedIds } from "../../../HomePageHelpers";
import Bin from "../Bin";

const mockRetrieveAllClick = jest.fn();
const mockEmptyBinClick = jest.fn();

describe("Bin", () => {
  test("renders correctly", () => {
    const { container } = render(
      <Bin
        trash={mockTrashTodos}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");

    expect(binButton).toBeInTheDocument();

    const randomIdsReplaced = replaceAllRandomGeneratedIds(container.innerHTML);

    expect(randomIdsReplaced).toMatchSnapshot();
  });

  test("menu renders correctly after clicking on bin button", async () => {
    const { container } = render(
      <Bin
        trash={mockTrashTodos}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");

    await userEvent.click(binButton);

    const retrieveAllMenuButton = screen.getByText(/retrieve all/i);
    const emptyBinMenuButton = screen.getByText(/empty bin/i);

    expect(retrieveAllMenuButton).toBeInTheDocument();
    expect(emptyBinMenuButton).toBeInTheDocument();

    const randomIdsReplaced = replaceAllRandomGeneratedIds(container.innerHTML);

    expect(randomIdsReplaced).toMatchSnapshot();
  });

  test("calls handleRetrieveAllClick on retrieve all button click", async () => {
    render(
      <Bin
        trash={mockTrashTodos}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");
    await userEvent.click(binButton);

    const retrieveAllMenuButton = screen.getByText(/retrieve all/i);
    await userEvent.click(retrieveAllMenuButton);

    expect(mockRetrieveAllClick).toHaveBeenCalled();
  });

  test("calls handleEmptyBinClick on empty bin button click", async () => {
    render(
      <Bin
        trash={mockTrashTodos}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");
    await userEvent.click(binButton);

    const emptyBinMenuButton = screen.getByText(/empty bin/i);
    await userEvent.click(emptyBinMenuButton);

    expect(mockEmptyBinClick).toHaveBeenCalled();
  });
});
