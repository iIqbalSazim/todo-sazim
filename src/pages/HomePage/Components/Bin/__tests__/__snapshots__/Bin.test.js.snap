// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bin menu renders correctly after clicking on bin button 1`] = `"<div class="mantine-Group-root mantine-qokdfp"><button class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-y99nq2" type="button" aria-haspopup="menu" aria-expanded="true" aria-controls="mantine-mockedId-dropdown" id="mantine-mockedId-target" data-expanded="true"><svg xmlns="http://www.w3.org/2000/svg" width="2rem" height="2rem" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1px" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-trash-filled"><path d="M20 6a1 1 0 0 1 .117 1.993l-.117 .007h-.081l-.919 11a3 3 0 0 1 -2.824 2.995l-.176 .005h-8c-1.598 0 -2.904 -1.249 -2.992 -2.75l-.005 -.167l-.923 -11.083h-.08a1 1 0 0 1 -.117 -1.993l.117 -.007h16z" fill="currentColor" stroke-width="0"></path><path d="M14 2a2 2 0 0 1 2 2a1 1 0 0 1 -1.993 .117l-.007 -.117h-4l-.007 .117a1 1 0 0 1 -1.993 -.117a2 2 0 0 1 1.85 -1.995l.15 -.005h4z" fill="currentColor" stroke-width="0"></path></svg></button><div class="mantine-Menu-dropdown mantine-1nbod7e" style="transition-property: opacity; transition-duration: 150ms; transition-timing-function: ease; opacity: 1; z-index: 300; top: 0px; left: -3.5px; width: 12.5rem;" aria-labelledby="mantine-mockedId-target" id="mantine-mockedId-dropdown" role="menu" tabindex="-1" data-position="left" aria-orientation="vertical"><div tabindex="-1" data-menu-dropdown="true" data-autofocus="true" style="outline: 0;"><button class="mantine-Menu-item mantine-eby8v3" type="button" tabindex="-1" role="menuitem" data-menu-item="true"><div class="mantine-1rnkhfs mantine-Menu-itemLabel">Retrieve all</div></button><button class="mantine-Menu-item mantine-eby8v3" type="button" tabindex="-1" role="menuitem" data-menu-item="true"><div class="mantine-1rnkhfs mantine-Menu-itemLabel">Empty Bin</div></button></div><div class="mantine-ndinzw mantine-Menu-arrow" style="width: 0.4375rem; height: 0.4375rem; transform: rotate(45deg); position: absolute; border-top-right-radius: 0rem; right: -0.21875rem; border-left-color: transparent; border-bottom-color: transparent; top: -0.21875px;"></div></div></div>"`;

exports[`Bin renders correctly 1`] = `"<div class="mantine-Group-root mantine-qokdfp"><button class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-y99nq2" type="button" aria-haspopup="menu" aria-expanded="false" aria-controls="mantine-mockedId-dropdown" id="mantine-mockedId-target"><svg xmlns="http://www.w3.org/2000/svg" width="2rem" height="2rem" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1px" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-trash-filled"><path d="M20 6a1 1 0 0 1 .117 1.993l-.117 .007h-.081l-.919 11a3 3 0 0 1 -2.824 2.995l-.176 .005h-8c-1.598 0 -2.904 -1.249 -2.992 -2.75l-.005 -.167l-.923 -11.083h-.08a1 1 0 0 1 -.117 -1.993l.117 -.007h16z" fill="currentColor" stroke-width="0"></path><path d="M14 2a2 2 0 0 1 2 2a1 1 0 0 1 -1.993 .117l-.007 -.117h-4l-.007 .117a1 1 0 0 1 -1.993 -.117a2 2 0 0 1 1.85 -1.995l.15 -.005h4z" fill="currentColor" stroke-width="0"></path></svg></button></div>"`;
